### PR TITLE
Rewrite `.encode('utf-8')` in some cases

### PR DESCRIPTION
**Pull request by mxr**
Attempts to address #138

### CODE REVIEW

File: pyupgrade.py [hunk 1]
```diff
@@ -1084,6 +1085,27 @@ def visit_Call(self, node):  # type: (ast.Call) -> None
...
         ):
             self.native_literals.add(_ast_to_offset(node))
+        elif (
+                isinstance(node.func, ast.Attribute) and
+                isinstance(node.func.value, ast.Str) and
+                node.func.attr == 'encode' and
+                not _starargs(node) and
+                len(node.args) == 1 and
+                isinstance(node.args[0], ast.Str) and
+                node.args[0].s.lower() == 'utf-8'
```
Comment (asottile) [inline pyupgrade.py]: `'utf8'` works too -- there's a function in either `codecs` or `encodings` that normalizes this value -- can probably reuse that here?

File: pyupgrade.py [hunk 2]
```diff
@@ -1084,6 +1085,27 @@ def visit_Call(self, node):  # type: (ast.Call) -> None
...
+        elif (
+                isinstance(node.func, ast.Attribute) and
+                isinstance(node.func.value, ast.Str) and
+                node.func.attr == 'encode' and
+                not _starargs(node) and
+                len(node.keywords) == 1 and
+                node.keywords[0].arg == 'encoding' and
+                isinstance(node.keywords[0].value, ast.Str) and
+                node.keywords[0].value.s.lower() == 'utf-8'
+        ):
```
Comment (asottile) [inline pyupgrade.py]: I wouldn't bother with this `if`, I don't think I've ever seen `.encode(encoding='UTF-8')`

File: pyupgrade.py [hunk 3]
```diff
@@ -1224,6 +1246,7 @@ def _fix_py3_plus(contents_text):
             visitor.six_type_ctx,
             visitor.six_with_metaclass,
             visitor.super_calls,
+            visitor.encode_calls,
```
Comment (asottile) [inline pyupgrade.py]: these are alphabetized, as are the attributes on `Py3PlusVisitor` (~roughly -- though grouped with their relevant private state)

File: tests/pyupgrade_test.py [hunk 4]
```diff
@@ -1448,6 +1448,30 @@ def test_fix_native_literals(s, expected):
...
+        ('"asd".encode(encoding="utf-8")', 'b"asd"'),
+        ('"asd".encode(encoding="UTF-8")', 'b"asd"'),
+
+        ('"☕️".encode("utf-8")', '"☕️".encode()'),
+        ('"☕️".encode("UTF-8")', '"☕️".encode()'),
+        ('"☕️".encode(encoding="utf-8")', '"☕️".encode()'),
+        ('"☕️".encode(encoding="UTF-8")', '"☕️".encode()'),
+
+        # the current version only works for string literals
+        ('x="asd"\nx.encode("utf-8")', 'x="asd"\nx.encode("utf-8")'),
```
Comment (asottile) [inline tests/pyupgrade_test.py]: the general trend I've done is to add a separate parametrized `_noop` test so it's easier to see

File: tests/pyupgrade_test.py [hunk 5]
```diff
@@ -1448,6 +1448,30 @@ def test_fix_native_literals(s, expected):
     assert _fix_py3_plus(s) == expected
 
 
+@pytest.mark.parametrize(
+    ('s', 'expected'),
+    (
+        ('"asd".encode("utf-8")', 'b"asd"'),
```
Comment (asottile) [inline tests/pyupgrade_test.py]: curious what does this does for:

```python
'a' 'b'.encode('UTF-8')

sys.stdout.buffer.write(
    'a'
    'b'.encode('UTF-8')
)
```

File: pyupgrade.py (hunk 2 above)
Comment (mxr) [reply pyupgrade.py]: OK

File: pyupgrade.py (hunk 1 above)
Comment (mxr) [reply pyupgrade.py]: Sure

Comment (mxr) [reply pyupgrade.py]: Hmm I found this one but it seems less than useful

```
In [5]: encodings.normalize_encoding('utf-8')
Out[5]: 'utf_8'

In [6]: encodings.normalize_encoding('utf8')
Out[6]: 'utf8'

In [7]: encodings.normalize_encoding('UTF8')
Out[7]: 'UTF8'
```

I wasn't able to find anything in codecs (from ipython autocomplete). Any advice?

File: pyupgrade.py (hunk 3 above)
Comment (mxr) [reply pyupgrade.py]: Sure

File: tests/pyupgrade_test.py (hunk 4 above)
Comment (mxr) [reply tests/pyupgrade_test.py]: sure

File: tests/pyupgrade_test.py (hunk 5 above)
Comment (mxr) [reply tests/pyupgrade_test.py]: Looks like my change causes mixing of byte & nonbyte literals. I could not handle this case at all or I could modify the predecessor to be a byte string if possible?

File: pyupgrade.py (hunk 1 above)
Comment (asottile) [reply pyupgrade.py]:
```pycon
>>> codecs.lookup('utf_8').name
'utf-8'
>>> codecs.lookup('utf8').name
'utf-8'
>>> codecs.lookup('UTF8').name
'utf-8'
```

File: tests/pyupgrade_test.py (hunk 5 above)
Comment (asottile) [reply tests/pyupgrade_test.py]: maybe let's not convert to bytestrings for now then, and just remove the encoding name -- I think that makes your change simpler / safer anyway 👍

Comment (mxr) [reply tests/pyupgrade_test.py]: suree

File: pyupgrade.py (hunk 1 above)
Comment (mxr) [reply pyupgrade.py]: ty!

File: tests/pyupgrade_test.py (hunk 5 above)
Comment (mxr) [reply tests/pyupgrade_test.py]: sure* `!sux mbp`

**Comment by mxr** [comment]
Can't request a review using GitHub for some reason but this is ready for another 👀

File: pyupgrade.py [hunk 6]
```diff
@@ -1092,6 +1104,13 @@ def generic_visit(self, node):
         super(FindPy3Plus, self).generic_visit(node)
 
 
+def _is_utf8_codec(encoding):
+    try:
+        return codecs.lookup(encoding).name == 'utf-8'
+    except LookupError:
+        return False
```
Comment (asottile) [inline pyupgrade.py]: I tend to put helpers _higher_ than their callers in the file, nbd though

File: tests/pyupgrade_test.py [hunk 7]
```diff
@@ -1448,6 +1448,39 @@ def test_fix_native_literals(s, expected):
...
+    ),
+)
+def test_fix_encode(s, expected):
+    assert _fix_py3_plus(s) == expected
+
+
+@pytest.mark.parametrize(
+    's',
+    (
+        # see https://git.io/fjRcs
```
Comment (asottile) [inline tests/pyupgrade_test.py]: let's just write out the comment, the link doesn't seem to work anyway
```python
# for now, pyupgrade is timid in this particular case
```

Comment (asottile) [reply tests/pyupgrade_test.py]: could also just move this down below the other comment about timidity (is that a word?)

Comment (mxr) [reply tests/pyupgrade_test.py]: sure

**Comment by asottile** [comment]
also looks like you need a test case for

```python
'foo'.encode('unknown-encoding')
```

File: pyupgrade.py (hunk 6 above)
Comment (mxr) [reply pyupgrade.py]: sure

**Comment by mxr** [comment]
Yea I didn't run coverage locally :( added one in the latest commit

File: pyupgrade.py [hunk 8]
```diff
@@ -981,6 +983,12 @@ def _is_six(self, node, names):
             node.attr in names
         )
 
+    def _is_utf8_codec(self, encoding):
```
Comment (asottile) [inline pyupgrade.py]: eheh not exactly what I meant -- since it doesn't make sense as a method (doesn't use `self`)

Comment (mxr) [reply pyupgrade.py]: I can move it

**Pull request merged**